File: minegame-broker/src/main/java/top/nipuru/minegame/broker/processor/PlayerPrivateChatBrokerProcessor.java
package top.nipuru.minegame.broker.processor;

import top.nipuru.minegame.broker.BrokerPlugin;
import top.nipuru.minegame.broker.game.GamePlayer;
import top.nipuru.minegame.broker.util.LeakBucketLimiter;
import top.nipuru.minegame.common.message.PlayerPrivateChatMessage;
import com.alipay.remoting.BizContext;
import com.alipay.remoting.rpc.protocol.SyncUserProcessor;
import net.afyer.afybroker.server.Broker;
import net.afyer.afybroker.server.proxy.BrokerPlayer;

public class PlayerPrivateChatBrokerProcessor extends SyncUserProcessor<PlayerPrivateChatMessage> {

    private final BrokerPlugin plugin;

    public PlayerPrivateChatBrokerProcessor(BrokerPlugin plugin) {
        this.plugin = plugin;
    }

    @Override
    public Object handleRequest(BizContext bizContext, PlayerPrivateChatMessage request) throws Exception {
        GamePlayer player = plugin.getPlayer(request.getSender().getPlayerId());
        if (player == null) {
            return PlayerPrivateChatMessage.FAILURE;
        }
        LeakBucketLimiter chatLimiter = player.getChatLimiter();
        if (chatLimiter.isLimit()) {
            return PlayerPrivateChatMessage.RATE_LIMIT;
        }
        BrokerPlayer receiver = Broker.getPlayer(request.getReceiver());
        if (receiver == null || receiver.getServer() == null) {
            return PlayerPrivateChatMessage.NOT_ONLINE;
        }
        return receiver.getServer().invokeSync(request);
    }

    @Override
    public String interest() {
        return PlayerPrivateChatMessage.class.getName();
    }
}

File: minegame-broker/src/main/kotlin/top/nipuru/minegame/broker/processor/PlayerPrivateChatBrokerProcessor.kt
package top.nipuru.minegame.broker.processor

import com.alipay.remoting.BizContext
import com.alipay.remoting.rpc.protocol.SyncUserProcessor
import net.afyer.afybroker.server.Broker
import top.nipuru.minegame.broker.playerList
import top.nipuru.minegame.common.message.PlayerPrivateChatMessage

class PlayerPrivateChatBrokerProcessor : SyncUserProcessor<PlayerPrivateChatMessage>() {

    override fun handleRequest(bizContext: BizContext, request: PlayerPrivateChatMessage): Any {
        val player = playerList.getPlayer(request.sender.playerId)
        val chatLimiter = player.chatLimiter
        if (chatLimiter.isLimit) {
            return PlayerPrivateChatMessage.RATE_LIMIT
        }
        val receiver = Broker.getPlayer(request.receiver)
        if (receiver == null || receiver.server == null) {
            return PlayerPrivateChatMessage.NOT_ONLINE
        }
        return receiver.server!!.invokeSync(request)
    }

    override fun interest(): String {
        return PlayerPrivateChatMessage::class.java.name
    }
}
